velox/functions/prestosql/MapZipWith.cpp

```diff
@@ -0,0 +1,289 @@
+        // Copy right key.
+        mergedKeys->copy(rightKeys.get(), i, rawRightIndices[i], 1);
+      } else {
+        // Copy left key.
+        mergedKeys->copy(leftKeys.get(), i, rawLeftIndices[i], 1);
+      }
+    }
+
+    std::vector<VectorPtr> lambdaArgs = {
+        mergedKeys, mergedLeftValues, mergedRightValues};
+
+    const SelectivityVector allElementRows(index);
+
+    VectorPtr mergedValues;
+
+    // Loop over lambda functions and apply these to (mergedKeys,
+    // mergedLeftValues, mergedRightValues). In most cases there will be only
+    // one function and the loop will run once.
+    auto it = args[2]->asUnchecked<FunctionVector>()->iterator(&rows);
+    while (auto entry = it.next()) {
+      SelectivityVector elementRows(index, false);
+      entry.rows->applyToSelected([&](auto row) {
+        elementRows.setValidRange(
+            rawNewOffsets[row], rawNewOffsets[row] + rawNewSizes[row], true);
+      });
+      elementRows.updateBounds();
+
+      BufferPtr wrapCapture;
+      if (entry.callable->hasCapture()) {
+        wrapCapture =
+            makeWrapCapture(*entry.rows, index, rawNewSizes, context.pool());
+      }
+
+      // Make sure already populated entries in newElements do not get
+      // overwritten.
+      VarSetter finalSelection(
+          context.mutableFinalSelection(), &allElementRows);
+      VarSetter isFinalSelection(context.mutableIsFinalSelection(), false);
+
+      entry.callable->apply(
+          elementRows,
+          allElementRows,
+          wrapCapture,
+          &context,
+          lambdaArgs,
+          &mergedValues);
+    }
+
+    auto localResult = std::make_shared<MapVector>(
+        context.pool(),
+        outputType,
+        newNulls,
+        rows.end(),
+        newOffsets,
+        newSizes,
+        mergedKeys,
+        mergedValues);
+    context.moveOrCopyResult(localResult, rows, result);
+  }
+
+  static std::vector<std::shared_ptr<exec::FunctionSignature>> signatures() {
+    // map(K, V1), map(K, V2), function(K, V1, V2, V3) -> map(K, V3)
+    return {exec::FunctionSignatureBuilder()
+                .typeVariable("K")
+                .typeVariable("V1")
+                .typeVariable("V2")
+                .typeVariable("V3")
+                .returnType("map(K,V3)")
+                .argumentType("map(K,V1)")
+                .argumentType("map(K,V2)")
+                .argumentType("function(K,V1,V2,V3)")
+                .build()};
+  }
+
+ private:
+  static void mergeKeys(
+      const VectorPtr& leftKeys,
+      const VectorPtr& rightKeys,
+      const std::vector<vector_size_t>& leftSorted,
+      const std::vector<vector_size_t>& rightSorted,
+      uint64_t* rawLeftNulls,
+      uint64_t* rawRightNulls,
+      vector_size_t* rawLeftIndices,
+      vector_size_t* rawRightIndices,
+      vector_size_t& index) {
+    const auto numLeft = leftSorted.size();
+    const auto numRight = rightSorted.size();
+
+    if (numLeft == 0) {
```

Comment:
Looks like these 2 `if` blocks can be omitted, if one side is empty, it will skip the `while` and jump to one of the `for` loops later.

Reply:
Good point. Let me change that.